crates/forge-runner/src/backtrace.rs

```diff
@@ -0,0 +1,226 @@
+    message: String,
+    contracts_data: &ContractsData,
+    encountered_errors: &[EncounteredError],
+) -> String {
+    if encountered_errors.is_empty() {
+        return message;
+    }
+
+    let is_backtrace_enabled = env::var(BACKTRACE_ENV).is_ok_and(|value| value == "1");
+    if !is_backtrace_enabled {
+        return format!(
+            "{message}\nnote: run with `{BACKTRACE_ENV}=1` environment variable to display a backtrace",
+        );
+    }
+
+    BacktraceContractFormatter::new(contracts_data, encountered_errors)
+        .map(|formatter| {
+            encountered_errors
+                .iter()
+                .filter_map(|error| formatter.get_backtrace(error.pc, error.class_hash))
+                .map(|backtrace| backtrace.to_string())
+                .collect::<Vec<_>>()
+                .join("\n")
+        })
+        .map_or_else(
+            |err| format!("{message}\nfailed to create backtrace: {err}"),
+            |backtraces| format!("{message}\n{backtraces}"),
+        )
+}
+
+struct ContractBacktraceData {
+    contract_name: String,
+    casm_debug_info_start_offsets: Vec<usize>,
+    coverage_annotations: CoverageAnnotationsV1,
+    profiler_annotations: ProfilerAnnotationsV1,
+}
+
+impl ContractBacktraceData {
+    fn new(class_hash: &ClassHash, contracts_data: &ContractsData) -> Result<Self> {
+        let contract_name = contracts_data
+            .get_contract_name(class_hash)
+            .context(format!(
+                "failed to get contract name for class hash: {class_hash}"
+            ))?
+            .clone();
+
+        let contract_artifacts = contracts_data
+            .get_artifacts(&contract_name)
+            .context(format!(
+                "failed to get artifacts for contract name: {contract_name}"
+            ))?;
+
+        let contract_class = serde_json::from_str::<ContractClass>(&contract_artifacts.sierra)?;
+
+        let sierra_debug_info = contract_class
+            .sierra_program_debug_info
+            .as_ref()
+            .context("debug info not found")?;
+
+        let VersionedCoverageAnnotations::V1(coverage_annotations) =
+            VersionedCoverageAnnotations::try_from_debug_info(sierra_debug_info).context(indoc! {
+                "perhaps the contract was compiled without the following entry in Scarb.toml under [profile.dev.cairo]:
+                unstable-add-statements-code-locations-debug-info = true
+
+                or scarb version is less than 2.8.0
+                "
+            })?;
+
+        let VersionedProfilerAnnotations::V1(profiler_annotations) =
+            VersionedProfilerAnnotations::try_from_debug_info(sierra_debug_info).context(indoc! {
+                "perhaps the contract was compiled without the following entry in Scarb.toml under [profile.dev.cairo]:
+                unstable-add-statements-functions-debug-info = true
+
+                or scarb version is less than 2.8.0
+                "
+            })?;
```

Comment:
Should we always try to add these annotations?

Reply:
Worth mentioning in the docs that this is needed as well. And let's try communicate with Scarb team if this is somehow implicitly resolvable.